StonehearthEditor/SuggestComboBoxCompanion.cs

```diff
@@ -0,0 +1,347 @@
+            // --> click-eventhandler 'SuggLbOnClick' is called
+            if (!_suggLb.Focused)
+                HideSuggBox();
+        }
+
+        protected void OnLocationChanged(object sender, EventArgs e)
+        {
+            return;
+            _suggLb.Top = comboBox.Top + comboBox.Height - 3;
+            _suggLb.Left = comboBox.Left + 3;
+        }
+
+        protected void OnSizeChanged(object sender, EventArgs e)
+        {
+            _suggLb.Width = comboBox.Width - 20;
+        }
+
+        private void SuggLbOnClick(object sender, EventArgs eventArgs)
+        {
+            comboBox.Text = _suggLb.Text;
+            comboBox.Focus();
+        }
+
+        private void HideSuggBox()
+        {
+            _suggLb.Visible = false;
+        }
+
+        protected void OnDropDown(object sender, EventArgs e)
+        {
+            HideSuggBox();
+        }
+
+        #region keystroke events
+
+        private bool skip = false;
+
+        /// <summary>
+        /// if the suggest-ListBox is visible some keystrokes
+        /// should behave in a custom way
+        /// </summary>
+        /// <param name="e"></param>
+        protected void OnPreviewKeyDown(object sender, PreviewKeyDownEventArgs e)
+        {
+            if (skip)
+            {
+                skip = false;
+                return;
+            }
+            else
+            {
+                skip = true;
+            }
+
+            if (!_suggLb.Visible)
+            {
+                return;
+            }
+            if (comboBox.DroppedDown)
+            {
+                comboBox.DroppedDown = false;
+            }
+
+            switch (e.KeyCode)
+            {
+                case Keys.Down:
+                    if (_suggLb.SelectedIndex < _suggBindingList.Count - 1)
+                        _suggLb.SelectedIndex++;
+                    return;
+                case Keys.Up:
+                    if (_suggLb.SelectedIndex > 0)
+                        _suggLb.SelectedIndex--;
+                    return;
+                case Keys.Tab:
+                case Keys.Enter:
+                    comboBox.Text = _suggLb.Text;
+                    comboBox.Select(0, comboBox.Text.Length);
+                    _suggLb.Visible = false;
+                    return;
+                case Keys.Escape:
+                    HideSuggBox();
+                    return;
+            }
+        }
+
+        private static readonly Keys[] KeysToHandle = new[]
+                    { Keys.Down, Keys.Up, Keys.Enter, Keys.Escape, Keys.Tab };
+
+
+        private void ComboBox_KeyDown(object sender, KeyEventArgs e)
+        {
+            // the keysstrokes of our interest should not be processed be base class:
+            if (_suggLb.Visible && KeysToHandle.Contains(e.KeyCode))
+            {
+                e.Handled = true;
+            }
+
+
+            /*if (!_suggLb.Visible)
```

Comment:
Delete, please, or explain why it's commented out then delete.

Reply:
👍